### PR TITLE
Inference-time prediction ensemble: average K best-val model outputs

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Inference-time prediction ensemble: average the **output predictions** from K independently-trained best-val checkpoints, then compute metrics on the averaged predictions. This sidesteps the catastrophic loss-basin problem that destroyed weight averaging (PR #554, val 6.87% → 92.85%), because we never touch the weights — we average in **output space**, which is always safe.

Expected gain: 1–3% improvement on `val_abupt_axis_mean_rel_l2_pct` from ensemble diversity. The K models already exist as W&B artifacts; there is zero training cost. This is K× inference only.

**Key insight:** Weight averaging (model soup) requires models to share a common loss basin (Wortsman et al. 2022, Ainsworth 2022). Our models are trained from independent random initializations → disconnected basins → weight averaging fails. Prediction averaging has no such constraint — diverse predictions always benefit from averaging when individual models are well-calibrated.

Reference: Lakshminarayanan et al., "Simple and Scalable Predictive Uncertainty Estimation using Deep Ensembles" (NeurIPS 2017); also standard practice on Kaggle and in production ML systems.

## Instructions

You will write a new standalone script `target/ensemble_eval.py` (do NOT modify `train.py`). The script:

1. **Loads K model checkpoints from W&B** — download the EMA checkpoint artifact for each run, instantiate the model from that run's config, and load the weights.

2. **Runs inference separately for each model** on the full val split and test split (match the eval settings from the original training run: `--eval-surface-points 65536 --eval-volume-points 65536`).

3. **Averages predictions in normalized space** — average `surface_preds` and `volume_preds` tensors across the K models BEFORE un-normalizing. Do not average in physical space.

4. **Un-normalizes the averaged predictions** and computes all 7-channel relative-L2 metrics:
   - `surface_pressure`, `wall_shear`, `volume_pressure`, `tau_x`, `tau_y`, `tau_z`
   - Aggregate: `abupt_axis_mean_rel_l2_pct` (mean of above 6 channels)

5. **Logs all metrics to W&B** under a new run, using `--wandb-group ensemble-inference-v1`.

### Checkpoint candidates (start with K=2, then K=3)

| Run ID | Agent | val_abupt (%) | Notes |
|--------|-------|----------------|-------|
| `9mm3sz7x` | askeladd | **6.8701** | SOTA — use this |
| `49aimdiz` | alphonse | ~6.87 | 2nd best — check W&B; if >7.0% skip it |
| 3rd best | any | TBD | Query W&B for next best val run |

To find the 3rd best checkpoint, run:
```python
import wandb
api = wandb.Api()
runs = api.runs("wandb-applied-ai-team/senpai-v1-drivaerml-ddp8",
                order="-summary_metrics.val_abupt_axis_mean_rel_l2_pct")
for r in runs[:10]:
    print(r.id, r.summary_metrics.get("val_abupt_axis_mean_rel_l2_pct"), r.state)
```

### Trials to run

**Trial 1 — K=2:** runs `9mm3sz7x` + `49aimdiz`
```bash
python target/ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name ensemble-k2
```

**Trial 2 — K=3:** add the 3rd-best checkpoint
```bash
python target/ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz <3rd-run-id> \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name ensemble-k3
```

**Trial 3 (optional) — K=4:** if K=3 still improves, add a 4th run

Report the val and test `abupt_axis_mean_rel_l2_pct` plus all 6 per-channel metrics for each trial.

### Implementation notes

- The EMA weights are what you want (not the raw model weights) — look for `ema_state_dict` in the W&B artifact or checkpoint file.
- Match the model architecture exactly from the run config: `model_layers`, `model_hidden_dim`, `model_heads`, `model_slices`, `pos_encoding_mode`, `use_qk_norm`, `rff_num_features`, `rff_init_sigmas`.
- You can load the dataset in eval-only mode — no training loop needed.
- Keep the script self-contained and runnable from the repo root with `WANDB_ENTITY` and `WANDB_PROJECT` env vars.

## Baseline

Current SOTA (run `9mm3sz7x`, askeladd, PR merged):

| Metric | Val (%) | Test (%) | AB-UPT target (%) |
|--------|---------|----------|-------------------|
| **abupt_axis_mean_rel_l2_pct** | **6.8701** | **8.1229** | — |
| surface_pressure | 3.823 | 4.074 | 3.82 |
| wall_shear | 7.637 | 7.264 | 7.29 |
| volume_pressure | 4.144 | 12.051 | 6.08 |
| tau_x | 6.501 | 6.177 | 5.35 |
| tau_y | 8.663 | 8.353 | 3.65 |
| tau_z | 10.266 | 9.266 | 3.63 |

W&B run: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/9mm3sz7x

**Beat target: val_abupt < 6.8701%**

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: nezuko — please acknowledge this assignment and begin the inference-time prediction ensemble experiment. Per the PR body, average the K best-val model checkpoint outputs at inference time.

Current SOTA to beat: val_abupt=6.9246% (PR #523). Report W&B run IDs and EP1 metrics once training starts.

**Comment by morganmcg1** [comment]
STUDENT nezuko:

## Results

**TL;DR:** Output-space ensembling works strongly. K=2 already beats SOTA by **5.20% relative** on val_abupt and **3.35%** on test_abupt — well above the PR's expected 1–3%. Diminishing-returns kick in around K=3, but K=5 still adds another 2pp of relative gain. Test-split improvement transfers cleanly: every channel improves at every K.

### Implementation

`target/ensemble_eval.py` (new standalone script, ~590 lines, train.py untouched). Pipeline:
1. Download each run's `best`-tagged W&B model artifact (cached on disk).
2. Reconstruct `SurfaceTransolver` from each run's `config.yaml` and load the saved EMA state-dict (saved checkpoints already have EMA weights baked in via `EMA.copy_to(base_model)` before `state_dict()`).
3. For each batch, forward through all K members in eval mode, average `surface_preds` and `volume_preds` **in normalized space**, then un-normalize via `TargetTransform`.
4. Accumulate `EvalAccumulator` exactly like `trainer_runtime.evaluate_split` so all 7 standard channels and case-level rel_L2 reproduce. K=1 re-eval matches BASELINE.md to <1e-4 (`val_abupt=6.8700` vs `6.8701`, `test_abupt=8.1228` vs `8.1229`).

bf16 autocast, batch=4, eval_chunk=1024, surface/volume=65536. Single A100; peak GPU 12.14 GB at K=5 (vs 9.77 GB at K=1).

### Run candidates and a deviation

PR-spec K=2 = `9mm3sz7x` + `49aimdiz`. For K=3 / K=4 / K=5 I picked the next-best by `full_val_abupt` (using `senpai-v1-drivaerml-ddp8` runs ordered ascending):

| Rank | Run ID | Agent | val_abupt (%) | full_val (%) | test (%) | Used in |
|---:|---|---|---:|---:|---:|---|
| 1 | `9mm3sz7x` | askeladd | 6.8701 | 6.8701 | 8.1229 | K=2,3,4,5 |
| 2 | `49aimdiz` | alphonse | 6.9404 | 6.8994 | 8.2972 | K=2,3,4,5 |
| 3 | `wyz68o8r` | thorfinn | 6.9246 | 6.9246 | 8.2355 | K=3,4,5 |
| 4 | `qqtdnlwq` | alphonse | 7.0063 | 7.0062 | 8.2921 | K=4,5 |
| 5 | `5o7jc7wi` | edward   | 7.0134 | 7.0134 | 8.3130 | K=5 |

**Deviation from suggested candidate list:** I initially included `loxzj4xq` (rank-3-ish by val) but its checkpoint is incompatible with the current `model.py` — it uses a multi-scale 3-band `StringSeparableEncoding` (`surface_string_sep.bands.0/1/2.{log_freq,phase}`) and an enlarged `project_surface_features.project.weight: [512, 196]` vs the standard `[512, 148]`. That's tanjiro's earlier custom architecture from a closed branch. I replaced it with `qqtdnlwq` (next compatible rank). Bug-fix-style note logged here for the advisor.

### Results — full table (re-evaluated under identical protocol)

All numbers from `ensemble_eval.py` with `--eval-surface-points 65536 --eval-volume-points 65536`, bf16 autocast, ensemble averaging in normalized space.

#### Validation split (34 cases, 7295 views, 1824 batches)

| K | abupt% | Δ vs K=1 | surface_p | wall_shear | vol_p | tau_x | tau_y | tau_z |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| **1 (SOTA)** | **6.8700** | — | 4.5153 | 7.7571 | 4.1441 | 6.7625 | 8.6628 | 10.2653 |
| 2 | 6.5126 | **−5.20%** | 4.2280 | 7.3825 | 3.8654 | 6.4267 | 8.2295 | 9.8136 |
| 3 | 6.3728 | **−7.24%** | 4.1250 | 7.2144 | 3.7813 | 6.2574 | 8.0685 | 9.6318 |
| 4 | 6.3343 | **−7.80%** | 4.0857 | 7.1778 | 3.7247 | 6.2109 | 8.0529 | 9.5971 |
| **5** | **6.2681** | **−8.76%** | **4.0216** | **7.1195** | **3.6755** | **6.1803** | **7.9482** | **9.5149** |

#### Test split (50 cases, 11091 views, 2773 batches)

| K | abupt% | Δ vs K=1 | surface_p | wall_shear | vol_p | tau_x | tau_y | tau_z |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| **1 (SOTA)** | **8.1228** | — | 4.1274 | 7.4534 | 12.0512 | 6.5662 | 8.3260 | 9.5430 |
| 2 | 7.8505 | **−3.35%** | 3.8869 | 7.1557 | 11.8899 | 6.2978 | 7.9820 | 9.1960 |
| 3 | 7.7049 | **−5.14%** | 3.7988 | 6.9849 | 11.7710 | 6.1277 | 7.8133 | 9.0139 |
| 4 | 7.6597 | **−5.70%** | 3.7559 | 6.9388 | 11.7175 | 6.0723 | 7.7848 | 8.9681 |
| **5** | **7.5813** | **−6.66%** | **3.7151** | **6.8999** | **11.5205** | **6.0551** | **7.7052** | **8.9108** |

#### vs AB-UPT public reference (test split)

| Channel | AB-UPT | K=1 SOTA | K=5 | K=5 status |
|---|---:|---:|---:|---|
| surface_pressure | 3.82 | 4.13 | **3.72** | **beats** ref |
| wall_shear (vec τ) | 7.29 | 7.45 | **6.90** | **beats** ref |
| volume_pressure | 6.08 | 12.05 | 11.52 | far worse — pre-existing weakness |
| tau_x | 5.35 | 6.57 | 6.06 | narrows gap (~13%) |
| tau_y | 3.65 | 8.33 | 7.71 | far worse — pre-existing weakness |
| tau_z | 3.63 | 9.54 | 8.91 | far worse — pre-existing weakness |

K=5 ensemble is the first model on this codebase to **beat the AB-UPT reference on surface_pressure and wall_shear simultaneously**. The remaining 3 channels are pre-existing baseline weaknesses that ensembling narrows but does not close.

### Resources

| K | wall-clock (val + test) | peak GPU | W&B run |
|---:|---|---:|---|
| 1 | 4.7 + 7.2 = 11.9 min | 9.77 GB | (re-eval, no-wandb) |
| 2 | 9.3 + 14.1 = 23.4 min | 11.98 GB | [`rm96kpkq`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/rm96kpkq) |
| 3 | 13.8 + 21.0 = 34.8 min | 12.04 GB | [`x741gq4u`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/x741gq4u) |
| 4 | 18.3 + 27.8 = 46.2 min | 12.09 GB | [`vzhn8zos`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/vzhn8zos) |
| 5 | 22.9 + 34.7 = 57.6 min | 12.14 GB | [`xpfzgke4`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/xpfzgke4) |

Memory grows ~0.05 GB per added model (KV-buffers and intermediate activations), wall-clock grows linearly. All ensembles fit comfortably on a single A100.

### Commands

```bash
# K=2 (PR-spec)
CUDA_VISIBLE_DEVICES=0 uv run python ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name nezuko/ensemble-k2-9mm3-49ai

# K=3
CUDA_VISIBLE_DEVICES=1 uv run python ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz wyz68o8r \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name nezuko/ensemble-k3

# K=4 (loxzj4xq replaced with qqtdnlwq, see deviation note)
CUDA_VISIBLE_DEVICES=2 uv run python ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz wyz68o8r qqtdnlwq \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name nezuko/ensemble-k4

# K=5
CUDA_VISIBLE_DEVICES=3 uv run python ensemble_eval.py \
  --run-ids 9mm3sz7x 49aimdiz wyz68o8r qqtdnlwq 5o7jc7wi \
  --split val test \
  --wandb-group ensemble-inference-v1 \
  --wandb-name nezuko/ensemble-k5
```

### What happened — analysis

**It worked, and clearly so.** Output-space averaging delivered 2× the predicted improvement at K=2 and kept compounding through K=5. Three observations:

1. **Sidesteps the loss-basin problem (as predicted in PR body).** PR #554 model-soup blew up val from 6.87% → 92.85% because the K independent runs sit in disconnected basins. Output-space averaging makes no basin assumption — it's just N random independent decoders mixed at the prediction layer, and the diversity benefit is real.

2. **Test-set generalization tracks val improvement.** Often ensemble val gains shrink on test (over-fitting to val-checkpoint selection). Here the test improvement is ~80% as large as val improvement (K=5: −8.76% val vs −6.66% test), and improvements are uniform across all 6 channels on both splits. Suggests the gain is from real model diversity rather than val-set memorization.

3. **Diminishing returns appear at K=3 but don't saturate.** Marginal gains: K=2→K=3 = 1.4pp val, K=3→K=4 = 0.4pp, K=4→K=5 = 0.7pp. The K=4→K=5 jump is bigger than K=3→K=4 because `5o7jc7wi` is from a different agent (edward) and was trained 13 epochs vs the 5–6 epochs of the others — adds genuine diversity rather than near-duplicate runs. Confirms that **member diversity > member quality** for the marginal contribution.

4. **AB-UPT-relevant channels move the most.** surface_pressure and wall_shear (the two channels where SOTA was already in the same league as AB-UPT) cross the AB-UPT threshold at K=5. The 3 weak channels (vol_p, tau_y, tau_z) narrow but the underlying single-model gap is too wide for an ensemble alone to close.

### Suggested follow-ups (not implemented)

- **Weighted ensembling with full_val-derived weights.** Currently uniform weights. Try weights ∝ `1/full_val_abupt` per member or temperature-scaled softmax of negative full_val. Cheap to test in this same script.
- **TTA (test-time augmentation) inside ensembling.** For each case, eval each member with K' different point-subsamples (e.g. 4 different stride offsets), then average. Stacks with member ensembling. Cost = K × K' × inference, but for K=5, K'=4 still fits in <4 hr on a single GPU.
- **Mass-weighted averaging by epoch / training time.** A run trained 13 epochs (`5o7jc7wi`) probably contributes more uncorrelated information than 6-epoch runs. Could weight by `n_epochs` or by member entropy.
- **Volume-pressure-targeted ensemble.** vol_p is the dominant test_abupt contribution (12.05 → ~9.6 needed to hit AB-UPT). Could train K members with a vol-loss weight schedule and ensemble those — likely to dent the 11.52 floor more than uniform diversity does.
- **Greedy ensemble selection.** With 5 candidates, all are useful. With 20+ candidates, greedy add-one-at-a-time on val_abupt would cap the run cost while preserving most of the gain (Caruana et al. 2004).

### Files changed

- `target/ensemble_eval.py` (new, ~590 lines)

**Comment by morganmcg1** [comment]
ADVISOR: Outstanding work, nezuko. This is the single largest single-round improvement in the programme.

**APPROVED FOR MERGE**

Results summary:
- K=5 ensemble val_abupt = **6.2681%** (prev SOTA 6.9246%, **−8.76% relative**)
- K=5 ensemble test_abupt = **7.5813%** (prev SOTA 8.2355%, **−6.66% relative**)
- First time surface_pressure (3.72%) AND wall_shear (6.90%) simultaneously beat AB-UPT reference on test
- K=5 beats AB-UPT on surface_pressure test (3.72% vs 3.82%)
- K=5 beats AB-UPT on wall_shear test (6.90% vs 7.29%)

The implementation is clean — standalone `ensemble_eval.py`, train.py untouched, single A100 at 12 GB peak VRAM. The architecture-mismatch guard (excluding loxzj4xq) is correct and well-reasoned.

**Policy decision:** The ensemble SOTA (6.2681% val / 7.5813% test) becomes the primary gate for all future ensemble-tier evaluation. Single-model training experiments continue to gate against 6.9246% val (single-model SOTA), since they produce single checkpoints. When a single-model training PR beats 6.9246%, we will also run ensemble_eval.py to check if the new checkpoint improves the K=5 ensemble.

**Suggested follow-ups I want to pursue:**
1. Weighted ensemble (1/val_abupt weights) — your suggestion
2. Greedy ensemble selection from a larger pool of candidates
3. Tau-targeted ensemble: select members specifically by tau_y/tau_z rather than overall abupt
4. K=6 or K=7 trial adding the next best single-model checkpoints

Merging now.
